crates/sui-core/src/signature_mpc/sign_state.rs

```diff
@@ -0,0 +1,122 @@
+
+#[derive(Clone)]
+pub(crate) struct SignState {
+    epoch: EpochId,
+    pub party_id: PartyID,
+    parties: HashSet<PartyID>,
+    aggregator_party_id: PartyID,
+    pub tiresias_public_parameters: DecryptionPublicParameters,
+    pub tiresias_key_share_decryption_key_share: SecretKeyShareSizedNumber,
+    pub messages: Option<Vec<Vec<u8>>>,
+    pub public_nonce_encrypted_partial_signature_and_proofs:
+        Option<Vec<PublicNonceEncryptedPartialSignatureAndProof<ProtocolContext>>>,
+    pub presigns: Option<Vec<DecentralizedPartyPresign>>,
+    pub decryption_shares:
+        HashMap<PartyID, Vec<(PaillierModulusSizedNumber, PaillierModulusSizedNumber)>>,
+    pub proofs: Option<HashMap<PartyID, Vec<(PartialDecryptionProof)>>>,
+    pub failed_messages_indices: Option<Vec<usize>>,
+    pub involved_parties: Vec<PartyID>,
+}
+
+impl SignState {
+    pub(crate) fn new(
+        tiresias_key_share_decryption_key_share: SecretKeyShareSizedNumber,
+        tiresias_public_parameters: DecryptionPublicParameters,
+        epoch: EpochId,
+        party_id: PartyID,
+        parties: HashSet<PartyID>,
+        session_id: SignatureMPCSessionID,
+    ) -> Self {
+        let aggregator_party_id = ((u64::from_be_bytes((&session_id.0[0..8]).try_into().unwrap())
+            % parties.len() as u64)
+            + 1) as PartyID;
+
+        Self {
+            epoch,
+            party_id,
+            parties,
+            aggregator_party_id,
+            tiresias_public_parameters,
+            messages: None,
+            public_nonce_encrypted_partial_signature_and_proofs: None,
+            decryption_shares: HashMap::new(),
+            tiresias_key_share_decryption_key_share,
+            presigns: None,
+            proofs: None,
+            failed_messages_indices: None,
+            involved_parties: Vec::new(),
+        }
+    }
+
+    pub(crate) fn set(
+        &mut self,
+        messages: Vec<Vec<u8>>,
+        public_nonce_encrypted_partial_signature_and_proofs: Vec<
+            PublicNonceEncryptedPartialSignatureAndProof<ProtocolContext>,
+        >,
+        presigns: Vec<DecentralizedPartyPresign>,
+    ) {
+        self.messages = Some(messages);
+        self.public_nonce_encrypted_partial_signature_and_proofs =
+            Some(public_nonce_encrypted_partial_signature_and_proofs);
+        self.presigns = Some(presigns);
+    }
+
+    pub(crate) fn insert_first_round(
+        &mut self,
+        party_id: PartyID,
+        message: SignMessage,
+    ) -> twopc_mpc_protocols::Result<()> {
+        match message {
+            SignMessage::DecryptionShares(shares) => {
+                let _ = self.decryption_shares.insert(party_id, shares);
+            }
+            SignMessage::Proofs { proofs, failed_messages_indices, involved_parties } => {
+                self.failed_messages_indices = Some(failed_messages_indices.clone());
+                self.involved_parties = involved_parties.clone();
```

Comment:
IIUC, these values are OVERRIDDEN with every incoming proof message.
I don't think we can guarantee each member sending in this value correctly.